File: go/knowledge/extract/subprocess.go
// SPDX-License-Identifier: Apache-2.0

package extract

import (
	"bytes"
	"context"
	"fmt"
	"os/exec"
	"sync"
	"time"
)

// SubprocessResult holds the output from a subprocess invocation.
type SubprocessResult struct {
	Stdout   []byte
	Stderr   string
	ExitCode int
}

// RunSubprocess executes a binary with the given arguments and optional
// stdin data. The context controls cancellation and the timeout is
// enforced as a hard wall-clock limit on top of the context deadline.
func RunSubprocess(ctx context.Context, binary string, args []string, stdin []byte, timeout time.Duration) (SubprocessResult, error) {
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, timeout)
		defer cancel()
	}

	cmd := exec.CommandContext(ctx, binary, args...)

	if len(stdin) > 0 {
		cmd.Stdin = bytes.NewReader(stdin)
	}

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			return SubprocessResult{}, fmt.Errorf("extract: running %s: %w", binary, err)
		}
	}

	return SubprocessResult{
		Stdout:   stdout.Bytes(),
		Stderr:   stderr.String(),
		ExitCode: exitCode,
	}, nil
}

// binaryCache caches the result of binary availability checks for a
// short period to avoid repeated filesystem lookups.
var binaryCache = struct {
	mu      sync.Mutex
	entries map[string]binaryCacheEntry
}{
	entries: make(map[string]binaryCacheEntry),
}

type binaryCacheEntry struct {
	available bool
	checkedAt time.Time
}

const binaryCacheTTL = 5 * time.Minute

// CheckBinaryAvailable returns true when the named binary can be found
// on PATH. Results are cached for five minutes.
func CheckBinaryAvailable(name string) bool {
	binaryCache.mu.Lock()
	defer binaryCache.mu.Unlock()

	if entry, ok := binaryCache.entries[name]; ok {
		if time.Since(entry.checkedAt) < binaryCacheTTL {
			return entry.available
		}
	}

	_, err := exec.LookPath(name)
	available := err == nil

	binaryCache.entries[name] = binaryCacheEntry{
		available: available,
		checkedAt: time.Now(),
	}

	return available
}

// ResetBinaryCache clears the binary availability cache. Intended for
// use in tests only.
func ResetBinaryCache() {
	binaryCache.mu.Lock()
	defer binaryCache.mu.Unlock()
	binaryCache.entries = make(map[string]binaryCacheEntry)
}

File: sdks/ts/memory/src/ingest/subprocess.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Shared subprocess runner utility for Phase 4 extractors. Provides
 * timeout-enforced subprocess execution and binary availability
 * checking with caching. Reused by image OCR (P4-1), audio
 * transcription (P4-2), and video processing (P4-3).
 */

import { type ChildProcess, execFile, spawn } from 'node:child_process'
import { promisify } from 'node:util'

const execFileAsync = promisify(execFile)

/** Result of a subprocess invocation. */
export type SubprocessResult = {
  readonly stdout: Buffer
  readonly stderr: string
  readonly exitCode: number
}

/** Configuration for subprocess invocation. */
export type SubprocessOptions = {
  readonly timeout?: number
  readonly cwd?: string
  readonly env?: Readonly<Record<string, string>>
  readonly signal?: AbortSignal
  readonly maxBuffer?: number
}

/** Default subprocess timeout in milliseconds (60 seconds). */
export const DEFAULT_SUBPROCESS_TIMEOUT_MS = 60_000

/** Default maximum buffer size for subprocess output (50 MB). */
const DEFAULT_MAX_BUFFER = 50 * 1024 * 1024

/**
 * Maximum number of simultaneous subprocess invocations. Prevents
 * resource exhaustion when many extractions run concurrently.
 */
export const MAX_CONCURRENT_SUBPROCESSES = 8

/**
 * Simple counting semaphore that limits concurrency. Callers await
 * acquire() before spawning a subprocess and call release() when done.
 */
const createSemaphore = (maxConcurrency: number) => {
  let current = 0
  const waiting: Array<() => void> = []

  const acquire = (): Promise<void> => {
    if (current < maxConcurrency) {
      current++
      return Promise.resolve()
    }
    return new Promise<void>((resolve) => {
      waiting.push(resolve)
    })
  }

  const release = (): void => {
    const next = waiting.shift()
    if (next !== undefined) {
      next()
    } else {
      current--
    }
  }

  return { acquire, release }
}

const subprocessSemaphore = createSemaphore(MAX_CONCURRENT_SUBPROCESSES)

/**
 * Executes a binary with the given arguments and optional stdin input.
 * Arguments are passed as an array — no shell expansion occurs.
 * Timeout is enforced via setTimeout + process.kill().
 *
 * Uses spawn() for stdin support since execFile's `input` option can
 * hang when combined with promisify.
 */
export const runSubprocess = async (
  binary: string,
  args: readonly string[],
  stdin?: Buffer,
  opts?: SubprocessOptions,
): Promise<SubprocessResult> => {
  await subprocessSemaphore.acquire()

  const timeout = opts?.timeout ?? DEFAULT_SUBPROCESS_TIMEOUT_MS
  const maxBuffer = opts?.maxBuffer ?? DEFAULT_MAX_BUFFER

  return new Promise<SubprocessResult>((resolve, reject) => {
    let child: ChildProcess
    let timedOut = false

    try {
      child = spawn(binary, [...args], {
        cwd: opts?.cwd,
        env: opts?.env ? { ...process.env, ...opts.env } : undefined,
        stdio: ['pipe', 'pipe', 'pipe'],
        signal: opts?.signal,
      })
    } catch (err: unknown) {
      subprocessSemaphore.release()
      reject(
        new Error(
          `ingest: subprocess "${binary}": ${err instanceof Error ? err.message : String(err)}`,
        ),
      )
      return
    }

    const timer = setTimeout(() => {
      timedOut = true
      child.kill('SIGKILL')
    }, timeout)

    const stdoutChunks: Buffer[] = []
    const stderrChunks: Buffer[] = []
    let stdoutBytes = 0
    let stderrBytes = 0

    child.stdout?.on('data', (chunk: Buffer) => {
      stdoutBytes += chunk.length
      if (stdoutBytes <= maxBuffer) {
        stdoutChunks.push(chunk)
      }
    })

    child.stderr?.on('data', (chunk: Buffer) => {
      stderrBytes += chunk.length
      if (stderrBytes <= maxBuffer) {
        stderrChunks.push(chunk)
      }
    })

    child.on('error', (err: Error) => {
      clearTimeout(timer)
      subprocessSemaphore.release()
      reject(new Error(`ingest: subprocess "${binary}": ${err.message}`))
    })

    child.on('close', (code: number | null) => {
      clearTimeout(timer)
      subprocessSemaphore.release()

      if (timedOut) {
        reject(new Error(`ingest: subprocess "${binary}" timed out after ${timeout}ms`))
        return
      }

      resolve({
        stdout: Buffer.concat(stdoutChunks),
        stderr: Buffer.concat(stderrChunks).toString('utf-8'),
        exitCode: code ?? 1,
      })
    })

    // Write stdin if provided, then close the stream.
    if (stdin !== undefined && child.stdin !== null) {
      child.stdin.end(stdin)
    } else if (child.stdin !== null) {
      child.stdin.end()
    }
  })
}

/** Cache entry for binary availability. */
type BinaryCacheEntry = {
  readonly available: boolean
  readonly expiresAt: number
}

/** Cache TTL in milliseconds (5 minutes). */
const BINARY_CACHE_TTL_MS = 5 * 60 * 1000

const binaryCache = new Map<string, BinaryCacheEntry>()

/**
 * Reports whether the named binary is present on the system PATH.
 * Results are cached for 5 minutes.
 */
export const checkBinaryAvailable = async (name: string): Promise<boolean> => {
  const cached = binaryCache.get(name)
  if (cached !== undefined && Date.now() < cached.expiresAt) {
    return cached.available
  }

  const available = await probeBinary(name)

  binaryCache.set(name, {
    available,
    expiresAt: Date.now() + BINARY_CACHE_TTL_MS,
  })

  return available
}

/**
 * Clears the binary availability cache. Intended for testing only.
 */
export const resetBinaryCache = (): void => {
  binaryCache.clear()
}

/**
 * Probes for a binary by attempting to run `which` (Unix) or
 * `where` (Windows).
 */
const probeBinary = async (name: string): Promise<boolean> => {
  const whichBinary = process.platform === 'win32' ? 'where' : 'which'
  try {
    await execFileAsync(whichBinary, [name], { timeout: 5000 })
    return true
  } catch {
    return false
  }
}
